pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,422 @@
+			}
+
+		}
+
+		if imagePullJob.Status.Desired == imagePullJob.Status.Succeeded {
+			succeeded = succeeded + 1
+		}
+	}
+
+	// 4. status
+	newStatus := &appsv1alpha1.ImageListPullJobStatus{
+		Desired:             int32(len(job.Spec.Images)),
+		Active:              active,
+		Completed:           completed,
+		Succeeded:           succeeded,
+		StartTime:           job.Status.StartTime,
+		FailedImageStatuses: failedImageStatuses,
+	}
+
+	now := metav1.NewTime(r.clock.Now())
+	if newStatus.StartTime == nil {
+		newStatus.StartTime = &now
+	}
+
+	if job.Spec.CompletionPolicy.Type != appsv1alpha1.Never && newStatus.Desired == newStatus.Completed {
+		newStatus.CompletionTime = &now
+	}
+	return newStatus
+}
+
+func (r *ReconcileImageListPullJob) syncImagePullJob(job *appsv1alpha1.ImageListPullJob, needToCreate, needToDelete []*appsv1alpha1.ImagePullJob) error {
+
+	//1. manage creating
+	var errs []error
+	if len(needToCreate) > 0 {
+		var createdNum int
+		var createdErr error
+
+		createdNum, createdErr = util.SlowStartBatch(len(needToCreate), slowStartInitialBatchSize, func(idx int) error {
+			imagePullJob := needToCreate[idx]
+			key := types.NamespacedName{Namespace: job.Namespace, Name: job.Name}.String()
+			scaleExpectations.ExpectScale(key, expectations.Create, imagePullJob.Spec.Image)
+			err := r.Create(context.TODO(), imagePullJob)
+			if err != nil {
+				scaleExpectations.ObserveScale(key, expectations.Create, imagePullJob.Spec.Image)
+				if !errors.IsTimeout(err) {
+					return fmt.Errorf("fail to create ImagePullJob (%s/%s): %s", imagePullJob.Namespace, imagePullJob.Spec.Image, err.Error())
+				}
+			}
+
+			return nil
+		})
+		if createdErr == nil {
+			r.recorder.Eventf(job, corev1.EventTypeNormal, "Successful create ImagePullJob", "Create %d ImagePullJob", createdNum)
+		} else {
+			errs = append(errs, createdErr)
+		}
+	}
+
+	//2. manage deleting
+	if len(needToDelete) > 0 {
+		var deleteErrs []error
+		for _, imagePullJob := range needToDelete {
+			key := types.NamespacedName{Namespace: job.Namespace, Name: job.Name}.String()
+			scaleExpectations.ExpectScale(key, expectations.Delete, imagePullJob.Spec.Image)
+			if err := r.Delete(context.TODO(), imagePullJob); err != nil {
+				scaleExpectations.ObserveScale(key, expectations.Delete, imagePullJob.Spec.Image)
+				deleteErrs = append(deleteErrs, fmt.Errorf("fail to delete ImagePullJob (%s/%s) for : %s", imagePullJob.Namespace, imagePullJob.Name, err))
+			}
+		}
+
+		if len(deleteErrs) > 0 {
+			errs = append(errs, deleteErrs...)
+		} else {
+			r.recorder.Eventf(job, corev1.EventTypeNormal, "Successful delete ImagePullJob", "Delete %d ImagePullJob", len(needToDelete))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (r *ReconcileImageListPullJob) filterImagesAndImagePullJobs(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) ([]string, []*appsv1alpha1.ImagePullJob) {
```

Comment:
Can you add UT to test the function filterImagesAndImagePullJobs, and contains add, delete image scenarios